kafka/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/AzureAuthenticationCallbackHandler.scala

```diff
@@ -0,0 +1,71 @@
+/**
+  * Copyright (c) 2013-present Snowplow Analytics Ltd.
+  * All rights reserved.
+  *
+  * This program is licensed to you under the Snowplow Community License Version 1.0,
+  * and you may not use this file except in compliance with the Snowplow Community License Version 1.0.
+  * You may obtain a copy of the Snowplow Community License Version 1.0 at https://docs.snowplow.io/community-license-1.0
+  */
+package com.snowplowanalytics.snowplow.collectors.scalastream
+package sinks
+
+import java.net.URI
+import java.{lang, util}
+
+import javax.security.auth.callback.Callback
+import javax.security.auth.callback.UnsupportedCallbackException
+import javax.security.auth.login.AppConfigurationEntry
+
+import org.apache.kafka.clients.producer.ProducerConfig
+import org.apache.kafka.common.security.auth.AuthenticateCallbackHandler
+import org.apache.kafka.common.security.oauthbearer.OAuthBearerToken
+import org.apache.kafka.common.security.oauthbearer.OAuthBearerTokenCallback
+
+import com.microsoft.azure.credentials.MSICredentials
+
+import com.nimbusds.jwt.JWTParser
+
+class AzureAuthenticationCallbackHandler extends AuthenticateCallbackHandler {
+
+  val credentials: MSICredentials = new MSICredentials()
+
+  var sbUri: String = ""
+
+  override def configure(
+    configs: util.Map[String, _],
+    saslMechanism: String,
+    jaasConfigEntries: util.List[AppConfigurationEntry]
+  ): Unit = {
+    val bootstrapServer = configs.get(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG).toString.replaceAll("\\[|\\]", "")
+    val uri             = URI.create("https://" + bootstrapServer)
+    this.sbUri = uri.getScheme + "://" + uri.getHost
+  }
+
+  override def handle(callbacks: Array[Callback]): Unit =
+    callbacks.foreach {
+      case callback: OAuthBearerTokenCallback =>
+        val token = getOAuthBearerToken()
```

Comment:
Is the callbacks array ever `!= 1` in length?  If yes, I would suggest to move `getOAuthBearerToken` above `callbacks.foreach`.  But if it is always array of length 1 then what you've done is good.

Reply:
Hmm, I think there should be at most one `OAuthBearerTokenCallback` callback but I will move it above just in case.

Reply:
Actually, if we move it above, `getOAuthBearerToken` will be called even if there is no `OAuthBearerTokenCallback` therefore it would be better to leave it as it is.